src/lib/storage/run_length_segment/run_length_segment_iterable.hpp

```diff
@@ -44,143 +43,192 @@ class RunLengthSegmentIterable : public PointAccessibleSegmentIterable<RunLength
+                              chunk_offset);
+    } else if (step_size < static_cast<int64_t>(linear_search_threshold)) {
+      const auto less_than_current = [&](const ChunkOffset offset) {
+        return offset < chunk_offset;
+      };
+      return std::find_if_not(end_positions->cbegin() + previous_end_position_index, end_positions->cend(),
+                              less_than_current);
+    } else {
+      return std::lower_bound(end_positions->cbegin() + previous_end_position_index, end_positions->cend(),
+                              chunk_offset);
+    }
+  }
+
  private:
   class Iterator : public BaseSegmentIterator<Iterator, SegmentPosition<T>> {
    public:
     using ValueType = T;
     using IterableType = RunLengthSegmentIterable<T>;
-    using ValueIterator = typename pmr_vector<T>::const_iterator;
-    using NullValueIterator = typename pmr_vector<bool>::const_iterator;
     using EndPositionIterator = typename pmr_vector<ChunkOffset>::const_iterator;
 
    public:
-    explicit Iterator(ValueIterator value_it, NullValueIterator null_value_it, EndPositionIterator end_position_it,
-                      EndPositionIterator end_position_begin_it, ChunkOffset chunk_offset)
-        : _value_it{std::move(value_it)},
-          _null_value_it{std::move(null_value_it)},
-          _end_position_it{std::move(end_position_it)},
-          _end_position_begin_it{std::move(end_position_begin_it)},
+    explicit Iterator(const std::shared_ptr<const pmr_vector<T>>& values,
+                      const std::shared_ptr<const pmr_vector<bool>>& null_values,
+                      const std::shared_ptr<const pmr_vector<ChunkOffset>>& end_positions,
+                      EndPositionIterator end_positions_it, ChunkOffset chunk_offset)
+        : _values{values},
+          _null_values{null_values},
+          _end_positions{end_positions},
+          _end_positions_it{std::move(end_positions_it)},
+          _end_positions_begin_it{_end_positions->cbegin()},
+          _linear_search_threshold{determine_linear_search_offset_distance_threshold(_end_positions)},
           _chunk_offset{chunk_offset} {}
 
    private:
     friend class boost::iterator_core_access;  // grants the boost::iterator_facade access to the private interface
 
     void increment() {
       ++_chunk_offset;
-
-      if (_chunk_offset > *_end_position_it) {
-        ++_value_it;
-        ++_null_value_it;
-        ++_end_position_it;
+      if (_chunk_offset > *_end_positions_it) {
+        ++_end_positions_it;
       }
     }
 
     void decrement() {
       --_chunk_offset;
-
-      // Make sure to only check the previous end position when we are not in the very first run.
-      if (_end_position_it != _end_position_begin_it && _chunk_offset <= *(_end_position_it - 1)) {
-        --_value_it;
-        --_null_value_it;
-        --_end_position_it;
+      if (_end_positions_it != _end_positions_begin_it && _chunk_offset <= *(_end_positions_it - 1)) {
```

Comment:
Do we need to check for the begin? If anything decrementing when we have reached the begin should enter a state in which dereferencing is invalid. UB would be fine, too.

Reply:
In any case, this invariant should hold:

```
auto it = begin();
--it;
++it;
*it == *begin();
```

Should we test for this?
```

Reply:
Wasn’t aware that this is allowed. Will take a look at it.

Reply:
I think survive "by luck" in most iterators as we use a variable of `ChunkOffset` to store the current position and that one will be decremented from the value of 0 to 2^32-1 and than incremented again. I don't know if this is not actually UB?

Do you know if decrementing from `begin()` is actually valid?

Reply:
I did not find the actual point in the specification (24.2.6 is somewhat related, but did not find it).
http://www.cplusplus.com/reference/iterator/BidirectionalIterator/ states for a bidirectional iterator and the expression `--a`: `Can be decremented (if a dereferenceable iterator value precedes it).`

Since there is not a dereferencable value that precedes `it` with `it = begin()`, I'd assume that `--it` is not valid.

Reply:
I see. So ++it can be out-of-bounds, but --it can't. Thanks for looking that up. Maybe add a DebugAssert and a comment pointing to the standard?

Reply:
Does it make sense to link to a draft? Or do we have access to an actual standard?

Reply:
Added
```
DebugAssert(_chunk_offset > 0, "An iterator pointing at the begin of a segment cannot be decremented, see "
                                     "§24.2.6 of http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4296.pdf "
                                     "(iterator can be decremented if a dereferencable iterator value precedes it).");
```

But I am not too happy with not having a really good cite of the actual C++ standard.

Reply:
The standard is not public. Don't get me started on that. Linking the draft is the normal modus operandi. I like https://eel.is/c++draft/ because it is easier to deep-link into a section.

Reply:
Changed to
```
// According to the C++ standard (see https://eel.is/c++draft/iterator.concept.bidir) a begin() iterator cannot
      // be decremented (an iterator can be decremented if a dereferencable iterator value precedes it).
      DebugAssert(_chunk_offset > 0, "An iterator pointing at the begin of a segment cannot be decremented");
```